native/membrane_videocompositor/src/elixir_bridge/scene_validation.rs

```diff
@@ -0,0 +1,456 @@
+use std::collections::{HashMap, HashSet};
+
+use super::elixir_structs::*;
+
+impl Object {
+    fn mentioned_names(&self) -> Vec<&Name> {
+        match self {
+            Object::Layout(layout) => layout.mentioned_names(),
+            Object::Texture(texture) => texture.mentioned_names(),
+            Object::Video(_) => Vec::new(),
+        }
+    }
+
+    fn previous_names(&self) -> Vec<&Name> {
+        match self {
+            Object::Layout(layout) => layout.previous_names(),
+            Object::Texture(texture) => texture.previous_names(),
+            Object::Video(_) => Vec::new(),
+        }
+    }
+}
+
+impl Texture {
+    fn mentioned_names(&self) -> Vec<&Name> {
+        let mut names = self.previous_names();
+
+        if let TextureOutputResolution::Name(name) = &self.resolution {
+            names.push(name);
+        }
+
+        names
+    }
+
+    fn previous_names(&self) -> Vec<&Name> {
+        vec![&self.input]
+    }
+}
+
+impl Layout {
+    fn mentioned_names(&self) -> Vec<&Name> {
+        let mut names = self.previous_names();
+
+        if let LayoutOutputResolution::Name(name) = &self.resolution {
+            names.push(name);
+        }
+
+        names
+    }
+
+    fn previous_names(&self) -> Vec<&Name> {
+        self.inputs.values().collect::<Vec<_>>()
+    }
+}
+
+impl Scene {
+    fn check_for_duplicate_pad_refs(
+        objects: &[(ObjectName, Object)],
+    ) -> Result<(), SceneParsingError> {
+        let mut input_pads = HashSet::new();
+
+        for (_, object) in objects {
+            if let Object::Video(InputVideo { input_pad }) = object {
+                if input_pads.contains(input_pad) {
+                    return Err(SceneParsingError::DuplicatePadReferences(input_pad.clone()));
+                }
+
+                input_pads.insert(input_pad);
+            }
+        }
```

Comment:
I believe you can use a more universal function to find duplicates, of smth like this:
```rust
fn fisrt_duplicate_element<T>(iter: T) -> Option<T>
where
    T: IntoIterator,
    T::Item: Eq + Hash,
{
    let mut uniq = HashSet::new();
    iter.into_iter().find(move |x| !uniq.insert(x))
}
```
Based on this [SO answer](https://stackoverflow.com/a/46767732).

Reply:
cool!